cocos2d/core/sprites/CCSpriteWebGLRenderCmd.js

```diff
@@ -24,21 +24,96 @@
 
 //Sprite's WebGL render command
 (function() {
+
     cc.Sprite.WebGLRenderCmd = function (renderable) {
         cc.Node.WebGLRenderCmd.call(this, renderable);
         this._needDraw = true;
 
         this._quad = new cc.V3F_C4B_T2F_Quad();
+        this._quadBufferView = new Uint32Array(this._quad.arrayBuffer);
         this._quadWebBuffer = cc._renderContext.createBuffer();
         this._quadDirty = true;
         this._dirty = false;
         this._recursiveDirty = false;
+        this._supportsBatching = true;
+        this._batchShader = cc.shaderCache.programForKey(cc.SHADER_POSITION_TEXTURECOLORALPHATEST_BATCHED);
+        this._matLocation = gl.getAttribLocation(this._batchShader._programObj, cc.ATTRIBUTE_NAME_MVMAT);
+        this._batchElementBuffer = null;
+        this._batchBuffer = null;
     };
 
     var proto = cc.Sprite.WebGLRenderCmd.prototype = Object.create(cc.Node.WebGLRenderCmd.prototype);
+
     proto.constructor = cc.Sprite.WebGLRenderCmd;
 
     proto.updateBlendFunc = function (blendFunc) {};
+    
+    proto.batchBufferPool = [];
+   
+    //creates webgl buffers and initializes their size to what is properly required for each sprite
+    proto.createBatchBuffer = function(numSprites)
+    {
+        var arrayBuffer = gl.createBuffer();
+        var elementBuffer = gl.createBuffer();
+
+        this.initBatchBuffers(arrayBuffer,elementBuffer,numSprites);
+
+        return {arrayBuffer: arrayBuffer, elementBuffer: elementBuffer, size: numSprites };
+    }
+
+    proto.initBatchBuffers = function(arrayBuffer, elementBuffer, numSprites)
+    {
+        gl.bindBuffer(gl.ARRAY_BUFFER, arrayBuffer);
+        gl.bufferData(gl.ARRAY_BUFFER, this.byteSizePerSprite * numSprites ,gl.DYNAMIC_DRAW);
+
+        gl.bindBuffer(gl.ELEMENT_ARRAY_BUFFER, elementBuffer);
+        gl.bufferData(gl.ELEMENT_ARRAY_BUFFER, this.indicesPerSprite * 2 * numSprites, gl.DYNAMIC_DRAW); //*2 because we use shorts for indices
+    }
+
+    //returns an object with {arrayBuffer, elementBuffer, size}, where size denotes how many sprites fit in the buffer (no need for bufferData if it's already big enough, bufferSubData enough)
+    proto.getBatchBuffer = function(numSprites)
+    {
+        var pool = this.batchBufferPool;
+        if(pool.length <=0)
+        {
+            return this.createBatchBuffer(numSprites);
+        }
+        else
+        {
+            var minBuf = null;  //we also track the smallest found buffer because that one will be re-initialized and returned if no fitting buffer can be found
+            var minSize = Number.MAX_VALUE; 
+            var minBufIndex = -1;
+            for(var i=pool.length-1;i>=0;--i)
+            {
+                var buf = pool[i];
+                if(buf.size >= numSprites)
+                {
+                    pool.removeByLastSwap(i);
```

Comment:
Oh no, it can. The bug in this piece of code is just that the code keeps making new buffers (unnecessarily) because it thinks that the pooled buffer is too small (since .size is never rewritten). That shouldn't cause the rendering to break, but it's wasteful for performance (but only maybe) and memory (until the old ones are garbage collected).

In case you're using bufferSubData differently in your code, note that it can only replace data if the total stuff you write into it remains in the size that you gave the buffer when you initialized it with bufferData(). bufferSubData can't resize the buffer.